lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -3229,7 +3229,7 @@ class ExprAvailabilityWalker : public ASTWalker {
       maybeDiagParameterizedExistentialErasure(EE, Where);
     }
     if (auto *CC = dyn_cast<ExplicitCastExpr>(E)) {
-      if (!isa<CoerceExpr>(CC) &&
+      if (!isa<CoerceExpr>(CC) && CC->getCastType() &&
```

Comment:
I took a quick look at this and I might have an idea about what is going on. Based on the stacktrace -`diagnoseStmtAvailability` walks into an `AbstractClosureExpr` because `ExprAvailabilityWalker` always returns `true` from `shouldWalkIntoClosure` _but_:

1. Result builders currently have multi-statement closure inference disabled due to some ordering issues related to one-way constraints;
2. In code completion mode we always pass `LeaveClosureBodiesUnchecked` to the constraint system;

I wonder if something like this might reproduce the crash:

```
import SwiftUI

struct S<T> {
}

extension S : ExpressibleByIntegerLiteral {
  typealias IntegerLiteralType = Int8

  init(integerLiteral: IntegerLiteralType) {
  }
}

struct MyView : View {
  var body: some View {
    Group {
      EmptyView()

      test {
        let x =  #^COMPLETE^# as S<String>
        print(x)
      }
    }
  }

  func test(_: () -> Void) -> some View {
    return EmptyView()
  }
}
```

Reply:
That example doesn’t reproduce the crash for me. AFAICT closures aren’t walked 
https://github.com/apple/swift/blob/ba42be5fcf2d663a589a3f5e4efe0ffcffb101fa/lib/Sema/MiscDiagnostics.cpp#L64-L65

I very much suspect that this issue will be resolved fundamentally when we get rid of `LeaveClosureBodiesUnchecked` and I’m wondering if it makes sense to hunt for a reproducing test case now since the fix seems pretty straight-forward.

Reply:
The check you pointed out would skip walking into some decls in the closures, not the closures themselves though. I think this issue needs to be addressed in a different way by adjusting the check in `ExprAvailabilityWalker`, otherwise it would just crash in some other place…

Reply:
I managed to find a reproducer that’s not code completion specific. With this reproducer at hand, I would argue that the fix is correct because AFAICT it appears correct that a `ExplicitCastExpr` has no `CastType` if that type is invalid. So we need to check if the type is valid (i.e. not null) before calling `hasParameterizedExistential`.

Reply:
How did an expression with incorrect cast end up getting past to a syntactic diagnostics?